scarb/tests/proc_macro_build.rs

```diff
@@ -0,0 +1,436 @@
+        .failure();
+}
+
+#[test]
+fn cannot_duplicate_macros_across_packages() {
+    let temp = TempDir::new().unwrap();
+    let t = temp.child("some");
+    CairoPluginProjectBuilder::default()
+        .lib_rs(indoc! {r#"
+        use cairo_lang_macro::{ProcMacroResult, TokenStream, attribute_macro};
+
+        #[attribute_macro]
+        pub fn hello(_attr: TokenStream, token_stream: TokenStream) -> ProcMacroResult {
+            ProcMacroResult::new(token_stream)
+        }
+
+        #[attribute_macro]
+        pub fn world(_attr: TokenStream, token_stream: TokenStream) -> ProcMacroResult {
+            ProcMacroResult::new(token_stream)
+        }
+        "#})
+        .build(&t);
+
+    let w = temp.child("other");
+    CairoPluginProjectBuilder::default()
+        .name("other")
+        .lib_rs(indoc! {r#"
+        use cairo_lang_macro::{ProcMacroResult, TokenStream, attribute_macro};
+
+        #[attribute_macro]
+        pub fn hello(_attr: TokenStream, token_stream: TokenStream) -> ProcMacroResult {
+            ProcMacroResult::new(token_stream)
+        }
+        "#})
+        .build(&w);
+
+    let project = temp.child("hello");
+    ProjectBuilder::start()
+        .name("hello")
+        .version("1.0.0")
+        .dep_starknet()
+        .dep("some", &t)
+        .dep("other", &w)
+        .lib_cairo(indoc! {r#"
+            #[hello]
+            #[world]
+            fn main() -> felt252 { 12 + 56 + 90 }
+        "#})
+        .build(&project);
+
+    Scarb::quick_snapbox()
+        .arg("build")
+        // Disable output from Cargo.
+        .env("CARGO_TERM_QUIET", "true")
+        .current_dir(&project)
+        .assert()
+        .failure()
+        .stdout_matches(indoc! {r#"
+            [..]Compiling some v1.0.0 ([..]Scarb.toml)
+            [..]Compiling other v1.0.0 ([..]Scarb.toml)
+            [..]Compiling hello v1.0.0 ([..]Scarb.toml)
+            error: duplicate expansions defined for procedural macros: hello (some v1.0.0 ([..]Scarb.toml) and other v1.0.0 ([..]Scarb.toml))
+        "#});
+}
+
+#[test]
+fn cannot_use_undefined_macro() {
+    let temp = TempDir::new().unwrap();
+    let t = temp.child("some");
+    CairoPluginProjectBuilder::default().build(&t);
+    let project = temp.child("hello");
+    ProjectBuilder::start()
+        .name("hello")
+        .version("1.0.0")
+        .dep_starknet()
+        .dep("some", &t)
+        .lib_cairo(indoc! {r#"
+            #[world]
+            fn main() -> felt252 { 12 + 56 + 90 }
+        "#})
+        .build(&project);
+    Scarb::quick_snapbox()
+        .arg("build")
+        // Disable output from Cargo.
+        .env("CARGO_TERM_QUIET", "true")
+        .current_dir(&project)
+        .assert()
+        .failure()
+        .stdout_matches(indoc! {r#"
+        [..]Compiling some v1.0.0 ([..]Scarb.toml)
+        [..]Compiling hello v1.0.0 ([..]Scarb.toml)
+        error: Plugin diagnostic: Unsupported attribute.
```

Comment:
Seems like a `proc_macro_diagnostics.rs` test? 🤔

Reply:
This diagnostic is not raised by our procedural macros system, but rather by attribute validations in the compiler itself. So I would not move it there, as it's not a diagnostic that we control.